File: leetcode/207.course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<pair<int, int>>& prerequisites) {
        
        unordered_map<int, vector<int>> graph;
        unordered_map<int, int> indegree;
        for (auto courses : prerequisites) {
            indegree[courses.first] += 1;
            graph[courses.second].push_back(courses.first);
        }
        
        queue<int> starts;
        for (auto i = 0; i < numCourses; ++i)
        {
            if (indegree.count(i) == 0)
            {
                starts.push(i);
            }
        }
        
        while (!starts.empty()) {
            int course = starts.front();
            starts.pop();
            
            vector<int> nodes = graph[course];
            for (auto i = 0; i < nodes.size(); ++i)
            {
                indegree[nodes[i]] -= 1;
                if (indegree[nodes[i]] == 0)
                {
                    starts.push(nodes[i]);
                    indegree.erase(nodes[i]);
                }
            }
        }
        
        return indegree.empty();
    }
};

File: leetcode/207.course-schedule.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        indegree = {i:0 for i in range(numCourses)}
        graph = {i:[] for i in range(numCourses)}
        take = 0
        
        for pair in prerequisites:
            indegree[pair[0]] += 1
            graph[pair[1]].append(pair[0])
                
        q = []
        for i in range(numCourses):
            if indegree[i] == 0:
                q.append(i)
                
        while(len(q) > 0):
            c = q.pop(0)
            take += 1
            
            for neighbor in graph[c]:
                if indegree.get(neighbor, 0) > 0:
                    indegree[neighbor] -= 1
                    if indegree[neighbor] == 0:
                        q.append(neighbor)
                        del indegree[neighbor]
                        
        return take == numCourses